book/src/rfc/const_generic.md

```diff
@@ -0,0 +1,265 @@
+ - the variable name `houses` in local scope
+ - start index of the `cvars` to represent the current access path
+ - the length of the `cvars` to represent size of the element targeted by the current access path
+
+So when it wants to update the elements inside the `houses`, it will need to reference the start index and the length of the `cvars` of the elements, so that it can narrow down the scope of the `cvars` to be updated.
+
+The following code is how it narrows down the scope for a field access or array access:
+
+```rust
+pub(crate) fn narrow(&self, start: usize, len: usize) -> Self {
+    match self {
+        VarOrRef::Var(var) => {
+            let cvars = var.range(start, len).to_vec();
+            VarOrRef::Var(Var::new(cvars, var.span))
+        }
+
+        //      old_start
+        //      |
+        //      v
+        // |----[-----------]-----| <-- var.cvars
+        //       <--------->
+        //         old_len
+        //
+        //
+        //          start
+        //          |
+        //          v
+        //      |---[-----]-|
+        //           <--->
+        //            len
+        //
+        VarOrRef::Ref {
+            var_name,
+            start: old_start,
+            len: old_len,
+        } => {
+            // ensure that the new range is contained in the older range
+            assert!(start < *old_len); // lower bound
+            assert!(start + len <= *old_len); // upper bound
+            assert!(len > 0); // empty range not allowed
+
+            Self::Ref {
+                var_name: var_name.clone(),
+                start: old_start + start,
+                len,
+            }
+        }
+    }
+}
+```
+
+This allows to find the range of the `cvars` belonging to the element targeted by the current access path under the local variable behind the `var_name`, which in this case is the `houses`. 
+
+But it doesn't keep track of the structure of the data. For example, when we are accessing `houses[1].rooms[2]`, it should be able to check if the access is within the array bounds. Although the current design can check if it is out of bound by checking access index against the total number of `cvars` for `houses`, it can't label which element access is out of bound as it could be either out of bound of `houses[1]` or the nested access `rooms[2]`.
+
+### Introduce ComputedExpr
+
+With the const variable as array size, we need to propagate its value during circuit generation to check if the access is within the bounds of the array. So we need a way to retain structure information of a variable that is currently represented by `VarOrRef`.
+
+We can add a new layer to the `VarOrRef` to keep track of the structure of the data. It is called `ComputedExpr`, as it is computed from `compute_expr` function, which resolves and passing around the values (`cvars`) via the `Expr`. The key difference from `VarOrRef` is `ComputedExpr` retains the structure of the underlying variables.
```

Comment:
I'm wondering if this is the kind of things that should prompt us to create an IR...

Reply:
I am sure `compute_expr` will be the focus when doing integration with circify ;)